node/src/components/consensus/era_supervisor.rs

```diff
@@ -1059,38 +1062,38 @@ where
             }
             ProtocolOutcome::ValidateConsensusValue {
                 sender,
-                consensus_value: candidate_block,
-                ancestor_values: ancestor_blocks,
+                proposed_block,
             } => {
                 if !self.era_supervisor.is_bonded(era_id) {
                     return Effects::new();
                 }
-                let proto_block = candidate_block.proto_block().clone();
-                let timestamp = candidate_block.timestamp();
-                let parent = candidate_block.parent().cloned();
-                let missing_evidence: Vec<PublicKey> = candidate_block
+                let missing_evidence: Vec<PublicKey> = proposed_block
+                    .value()
                     .accusations()
                     .iter()
                     .filter(|pub_key| !self.has_evidence(era_id, (*pub_key).clone()))
                     .cloned()
                     .collect();
                 self.era_mut(era_id)
-                    .add_candidate(candidate_block, missing_evidence.clone());
-                let proto_block_deploys_set: BTreeSet<DeployHash> =
-                    proto_block.deploys_and_transfers_iter().cloned().collect();
-                for ancestor_block in ancestor_blocks {
-                    let ancestor_proto_block = ancestor_block.proto_block();
-                    for deploy in ancestor_proto_block.deploys_and_transfers_iter() {
-                        if proto_block_deploys_set.contains(deploy) {
-                            return self.resolve_validity(
-                                era_id,
-                                sender,
-                                proto_block,
-                                parent,
-                                false,
-                            );
-                        }
-                    }
+                    .add_block(proposed_block.clone(), missing_evidence.clone());
+                let block_deploys_set: BTreeSet<DeployHash> = proposed_block
+                    .value()
+                    .deploys_and_transfers_iter()
+                    .cloned()
+                    .collect();
+                if proposed_block
+                    .context()
+                    .ancestor_values()
+                    .iter()
+                    .flat_map(|ancestor| ancestor.deploys_and_transfers_iter())
+                    .any(|deploy| block_deploys_set.contains(deploy))
+                {
+                    return self.resolve_validity(ResolveValidity {
+                        era_id,
+                        sender,
+                        proposed_block,
+                        valid: false,
+                    });
```

Comment:
Can we extract this to a method (or at least add a comment) saying this is a replay attack validation?

Reply:
Done in https://github.com/casper-network/casper-node/pull/1290/commits/794ed59f41d5773bdfe4357c8ae592e4f9670aaa.